File: react-native/ios/CoinbaseWalletSDKModule.swift
import ExpoModulesCore
import CoinbaseWalletSDK
import Foundation

public class CoinbaseWalletSDKModule: Module {

    public func definition() -> ModuleDefinition {

        Name("CoinbaseWalletSDK")

        Function("configure") { (params: ConfigParamsRecord) in
            guard !CoinbaseWalletSDK.isConfigured else {
                return
            }

            let host: URL
            if let hostURLStr = params.hostURL {
                host = URL(string: hostURLStr)!
            } else {
                host = URL(string: "https://wallet.coinbase.com/wsegue")!
            }

            CoinbaseWalletSDK.configure(
                host: host,
                callback: URL(string: params.callbackURL)!
            )
            CoinbaseWalletSDK.appendVersionTag("rn")
        }

        AsyncFunction("initiateHandshake") { (params: HandshakeParamsRecord, promise: Promise) in
            let actions: [Action] = params.initialActions.map { $0.asAction }

            DispatchQueue.main.async {
                CoinbaseWalletSDK.shared.initiateHandshake(initialActions: actions) { result, account in
                    switch result {
                    case .success(let response):
                        let results: [ActionResultRecord.Dict] = response.content.map { $0.asRecord }
                        let accountRecord = account?.asRecord
                        promise.resolve([results, accountRecord])
                    case .failure(let error):
                        promise.reject("handshake-error", "\(error)");
                    }
                }
            }
        }

        AsyncFunction("makeRequest") { (params: RequestParamsRecord, promise: Promise) in
            let requestActions: [Action] = params.actions.map { $0.asAction }

            let requestAccount: Account?
            if let account = params.account {
                requestAccount = Account(
                    chain: account.chain,
                    networkId: UInt(account.networkId),
                    address: account.address
                )
            } else {
                requestAccount = nil
            }
            
            DispatchQueue.main.async {
                CoinbaseWalletSDK.shared.makeRequest(
                    Request(actions: requestActions, account: requestAccount)
                ) { result in
                    switch result {
                    case .success(let response):
                        let results: [ActionResultRecord.Dict] = response.content.map { $0.asRecord }
                        promise.resolve(results)
                    case .failure(let error):
                        promise.reject("request-error", "\(error)")
                    }
                }
            }
        }

        Function("handleResponse") { (url: String) -> Bool in
            guard CoinbaseWalletSDK.isConfigured else {
                return false
            }

            let responseURL = URL(string: url)!
            if (try? CoinbaseWalletSDK.shared.handleResponse(responseURL)) == true {
                return true
            }

            return false
        }

        AsyncFunction("isCoinbaseWalletInstalled") { (promise: Promise) in
            DispatchQueue.main.async {
                promise.resolve(CoinbaseWalletSDK.isCoinbaseWalletInstalled())
            }
        }
        
        AsyncFunction("getCoinbaseWalletMWPVersion") { (promise: Promise) in
            DispatchQueue.main.async {
                promise.resolve(CoinbaseWalletSDK.getCoinbaseWalletMWPVersion())
            }
        }

        Function("isConnected") { () -> Bool in
            return CoinbaseWalletSDK.shared.isConnected()
        }

        Function("resetSession") {
            CoinbaseWalletSDK.shared.resetSession()
        }
    }
}

File: react-native/android/src/main/java/expo/modules/coinbasewalletsdkexpo/CoinbaseWalletSDKModule.kt
package expo.modules.coinbasewalletsdkexpo

import android.net.Uri
import android.util.Log
import com.coinbase.android.nativesdk.CoinbaseWalletSDK
import com.coinbase.android.nativesdk.message.request.Account
import com.coinbase.android.nativesdk.message.request.Action
import com.coinbase.android.nativesdk.message.request.RequestContent
import expo.modules.coinbasewalletsdkexpo.records.AccountRecord
import expo.modules.coinbasewalletsdkexpo.records.ActionRecord
import expo.modules.coinbasewalletsdkexpo.records.ActionResultRecord
import expo.modules.coinbasewalletsdkexpo.records.ConfigParamsRecord
import expo.modules.coinbasewalletsdkexpo.records.HandshakeParamsRecord
import expo.modules.coinbasewalletsdkexpo.records.RequestParamsRecord
import expo.modules.coinbasewalletsdkexpo.records.asAction
import expo.modules.coinbasewalletsdkexpo.records.asRecord
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class CoinbaseWalletSDKModule : Module() {
    private var sdk: CoinbaseWalletSDK? = null

    override fun definition() = ModuleDefinition {

        Name("CoinbaseWalletSDK")

        Function("configure") { params: ConfigParamsRecord ->
            try {
                IntentLauncher.onResult = { uri -> sdk?.handleResponse(uri) }

                val context = requireNotNull(appContext.reactContext?.applicationContext) {
                    "CoinbaseWalletSDK: Application context must not be null"
                }

                sdk = CoinbaseWalletSDK(
                    domain = Uri.parse(params.callbackURL),
                    appContext = context,
                    hostPackageName = params.hostPackageName ?: "org.toshi",
                    openIntent = { IntentLauncher.launcher?.launch(it) }
                )
                sdk?.appendVersionTag("rn")
            } catch (e: Exception) {
                Log.e("CoinbaseWalletSDK", "Configuration error", e)
            }
        }

        AsyncFunction("initiateHandshake") { params: HandshakeParamsRecord, promise: Promise ->
            val sdk = sdk
            if (sdk == null) {
                promise.reject("configure-error", "configure must be called before handshake can be initiated", null)
                return@AsyncFunction
            }

            val handshakeActions = params.initialActions.map { it.asAction }

            sdk.initiateHandshake(handshakeActions) { result, account ->
                result
                    .onSuccess { responses ->
                        val results: List<ActionResultRecord> = responses.map { it.asRecord }
                        val accountRecord = account?.asRecord
                        promise.resolve(listOf(results, accountRecord))
                    }
                    .onFailure { error ->
                        Log.e("CoinbaseWalletSDK", "Handshake error", error)
                        promise.reject("handshake-error", error.message, error)
                    }
            }
        }

        AsyncFunction("makeRequest") { params: RequestParamsRecord, promise: Promise ->
            val sdk = sdk
            if (sdk == null) {
                promise.reject("configure-error", "configure must be called before request can be initiated", null)
                return@AsyncFunction
            }

            val requestActions = params.actions.map { it.asAction }

            val requestAccount = params.account?.let { record ->
                Account(
                    chain = record.chain,
                    networkId = record.networkId.toLong(),
                    address = record.address
                )
            }

            val request = RequestContent.Request(actions = requestActions, account = requestAccount)
            sdk.makeRequest(request) { result ->
                result
                    .onSuccess { responses ->
                        val results: List<ActionResultRecord> = responses.map { it.asRecord }
                        promise.resolve(results)
                    }
                    .onFailure { error ->
                        Log.e("CoinbaseWalletSDK", "Request error", error)
                        promise.reject("request-error", error.message, error)
                    }
            }
        }

        Function("handleResponse") { url: String ->
            val responseURL = Uri.parse(url)

            try {
                return@Function sdk?.handleResponse(responseURL) ?: false
            } catch (error: Exception) {
                return@Function false
            }
        }

        Function("isCoinbaseWalletInstalled") {
            return@Function sdk?.isCoinbaseWalletInstalled ?: false
        }

        Function("isConnected") {
            return@Function sdk?.isConnected ?: false
        }

        Function("resetSession") {
            sdk?.resetSession()
        }
    }
}
